### PR TITLE
Add multiple namespace support 

### LINKED ISSUE #37: Add multiple namespace support
Add multiple namespace support 

## Expected Behaviour

Function names in faas-idler are not namespace aware. As other components now support multiple namespaces it is required for faas-idler to be namespace aware. 

## Current Behaviour

Does not support multiple namespace.

## Possible Solution

Implement a similar logic as SDK-Connector to make this namespace awar.

## Steps to Reproduce (for bugs)

1.
2.
3.
4.

## Your Environment

* Docker version `docker version` (e.g. Docker 17.0.05 ):

* Are you using Docker Swarm or Kubernetes (FaaS-netes)?

* Operating System and version (e.g. Linux, Windows, MacOS):

* Link to your project or a code example to reproduce issue:

* Please also follow the [troubleshooting guide](https://github.com/openfaas/faas/blob/master/guide/troubleshooting.md) and paste in any other diagnostic information you have:

### PR DESCRIPTION
**Pull request by viveksyngh**
## Description

Fixes #37 

## How Has This Been Tested?

I have tested this on my local system with k3d. 

1. Deployed OpenFaaS with multiple namespace (openfaas-fn, fn)
2. Deployed same functions (nodeinfo, figlet) in both name space 
3. Invoked the functions in both namespace 
```
2020/08/31 10:54:26 Gateway version: 0.18.18, SHA: 11ff356ccea32b89b9da556c18305cd7c5ea9ba5
2020/08/31 10:54:26 dry_run: false
gateway_url: http://gateway.openfaas:8080/
inactivity_duration: 5m0s
reconcile_interval: 2m0s
2020/08/31 10:54:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22nodeinfo%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 10:54:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22figlet%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 10:54:26 nodeinfo	idle
2020/08/31 10:54:26 figlet	idle
2020/08/31 10:54:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22nodeinfo.fn%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 10:54:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22figlet.fn%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 10:54:26 nodeinfo.fn	idle
2020/08/31 10:54:26 figlet.fn	idle
2020/08/31 10:54:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22nodeinfo.openfaas-fn%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 10:54:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22figlet.openfaas-fn%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 10:54:26 nodeinfo.openfaas-fn	idle
2020/08/31 10:56:26
2020/08/31 10:56:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22nodeinfo%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 10:56:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22figlet%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 10:56:26 nodeinfo	idle
2020/08/31 10:56:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22nodeinfo.fn%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 10:56:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22figlet.fn%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 10:56:26 nodeinfo.fn	idle
2020/08/31 10:56:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22nodeinfo.openfaas-fn%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 10:56:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22figlet.openfaas-fn%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 10:56:26 nodeinfo.openfaas-fn	idle
2020/08/31 10:58:26
2020/08/31 10:58:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22nodeinfo%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 10:58:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22figlet%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 10:58:26 nodeinfo	idle
2020/08/31 10:58:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22nodeinfo.fn%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 10:58:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22figlet.fn%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 10:58:26 nodeinfo.fn	idle
2020/08/31 10:58:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22nodeinfo.openfaas-fn%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 10:58:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22figlet.openfaas-fn%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 10:58:26 nodeinfo.openfaas-fn	idle
2020/08/31 11:00:26
2020/08/31 11:00:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22nodeinfo%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 11:00:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22figlet%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 11:00:26 nodeinfo	idle
2020/08/31 11:00:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22nodeinfo.fn%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 11:00:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22figlet.fn%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 11:00:26 nodeinfo.fn	idle
2020/08/31 11:00:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22nodeinfo.openfaas-fn%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 11:00:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22figlet.openfaas-fn%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 11:00:26 nodeinfo.openfaas-fn	idle
2020/08/31 11:02:26
2020/08/31 11:02:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22nodeinfo%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 11:02:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22figlet%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 11:02:26 nodeinfo	idle
2020/08/31 11:02:26 figlet	idle
2020/08/31 11:02:26 Scale figlet 202 0
2020/08/31 11:02:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22nodeinfo.fn%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 11:02:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22figlet.fn%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 11:02:26 nodeinfo.fn	idle
2020/08/31 11:02:26 figlet.fn	idle
2020/08/31 11:02:26 Scale figlet 202 0
2020/08/31 11:02:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22nodeinfo.openfaas-fn%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 11:02:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22figlet.openfaas-fn%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 11:02:26 nodeinfo.openfaas-fn	idle
2020/08/31 11:04:26
2020/08/31 11:04:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22nodeinfo%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 11:04:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22figlet%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 11:04:26 nodeinfo	idle
2020/08/31 11:04:26 figlet	idle
2020/08/31 11:04:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22nodeinfo.fn%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 11:04:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22figlet.fn%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 11:04:26 nodeinfo.fn	idle
2020/08/31 11:04:26 figlet.fn	idle
2020/08/31 11:04:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22nodeinfo.openfaas-fn%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 11:04:26 Query: sum%28rate%28gateway_function_invocation_total%7Bfunction_name%3D%22figlet.openfaas-fn%22%2C+code%3D~%22.%2A%22%7D%5B5m%5D%29%29+by+%28code%2C+function_name%29
2020/08/31 11:04:26 nodeinfo.openfaas-fn	idle
2020/08/31 11:06:26
```
4. Invoked the function (figlet) in only one namespace 

Function in other namespace was at 0 replicas while current namespace function has 1 repliacs 

## Checklist:

I have:

- [ ] updated the documentation if required
- [x] read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] signed-off my commits with `git commit -s`
- [ ] added unit tests

### CODE REVIEW

**Comment by goggi** [comment]
Great work! Any ETA on the release for the fix?

**Comment by alexellis** [comment]
This is already on the public roadmap, but there is no ETA for the open-source features.

Note that we do prioritise the roadmap and feature requests for [Premium Subscription clients](https://www.openfaas.com/support), if that's of interest, then take a look?

**Comment by alexellis** [comment]
(@viveksyngh, out of interest, are any changes required for this PR?)

File: main.go [hunk 1]
```diff
@@ -100,32 +134,35 @@ func readFile(path string) (string, error) {
...
-			duration))
+		//Deriving the function name for multiple namespace support
+		sep := ""
+		if len(namespace) > 0 {
+			sep = "."
+		}
+		functionName := fmt.Sprintf("%s%s%s", function.Name, sep, namespace)
 
-		log.Printf("Querying: %s\n", querySt)
+		querySt := url.QueryEscape(`sum(rate(gateway_function_invocation_total{function_name="` + functionName + `", code=~".*"}[` + duration + `])) by (code, function_name)`)
```
Comment (alexellis) [inline main.go:152]: This shouldn't have changed as part of the PR? Can you explain the change?

**Review by alexellis — CHANGES_REQUESTED** [review]
I noticed a change that I wasn't expecting

File: main.go (hunk 1 above)
Comment (viveksyngh) [reply main.go]: I have removed that changes .. it go introduced while, I was testing a different PR.

**Comment by viveksyngh** [comment]
I have removed the unexpected change.

**Review by alexellis — APPROVED** [review]
Approved

**Pull request merged**